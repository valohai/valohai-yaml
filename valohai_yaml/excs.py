from typing import List, Union

import jsonschema


class ValidationError(ValueError):
    """Generic validation error."""


class ValidationErrors(ValidationError):
    """Wrapper for multiple validation errors."""

    def __init__(self, errors: List[Union[str, jsonschema.ValidationError]]) -> None:
        self.errors = errors
        super().__init__(
            '%d errors: %s' % (
                len(errors),
                ', '.join(getattr(e, 'message', e) for e in self.errors)
            )
        )

    def __iter__(self):  # noqa: ANN204
        """Iterate over the errors contained within."""
        return iter(self.errors)
