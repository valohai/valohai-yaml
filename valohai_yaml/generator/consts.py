DEFAULT_DOCKER_IMAGE = "python:3.7"
