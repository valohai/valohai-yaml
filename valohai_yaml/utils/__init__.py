from yaml import safe_load


def read_yaml(yaml):
    if isinstance(yaml, (dict, list)):  # Smells already parsed
        return yaml
    if isinstance(yaml, bytes):
        yaml = yaml.decode('utf-8')
    return safe_load(yaml)  # can be a stream or a string


def listify(value):
    """
    Wrap the given value into a list, with the below provisions:

    * If the value is a list or a tuple, it's coerced into a new list.
    * If the value is None, an empty list is returned.
    * Otherwise, a single-element list is returned, containing the value.

    :param value: A value.
    :return: a list!
    :rtype: list
    """
    if value is None:
        return []
    if isinstance(value, (list, tuple)):
        return list(value)
    return [value]
