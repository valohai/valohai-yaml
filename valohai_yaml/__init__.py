from valohai_yaml.excs import ValidationErrors
from valohai_yaml.parsing import parse
from valohai_yaml.validation import validate

__version__ = '0.23.0'

__all__ = [
    'ValidationErrors',
    '__version__',
    'parse',
    'validate',
]
