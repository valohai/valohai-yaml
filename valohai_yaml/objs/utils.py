from collections import OrderedDict
from enum import Enum
from typing import TYPE_CHECKING, Any, Callable, Iterable, List, Optional, Type, TypeVar
from typing import OrderedDict as OrderedDictType

from valohai_yaml.types import SerializedDict

if TYPE_CHECKING:
    from valohai_yaml.objs.base import Item

TItem = TypeVar("TItem", bound="Item")
T = TypeVar("T")


def consume_array_of(
    source: SerializedDict,
    key: str,
    type: Type[TItem],
) -> List[TItem]:
    return [type.parse(datum) for datum in source.pop(key, ())]


def check_type_and_listify(
    source: Optional[Iterable[Any]],
    type: Type[T],
    parse: Optional[Callable[[Any], T]] = None,
) -> List[T]:
    """
    Check that all items in the `source` iterable are of the type `type`, return a list.

    If `parse` is given, and the item is not of the type, that function is called to parse it to one.

    """
    if source is None:
        return []
    out = []
    for item in source:
        if not isinstance(item, type):
            if not parse:
                raise TypeError(f"{item} not a {type}")
            item = parse(item)
            assert isinstance(item, type)  # Make sure `parse` was up to spec
        out.append(item)
    return out


def check_type_and_dictify(
    source: Optional[Iterable[Any]],
    type: Type[T],
    attr: str,
) -> OrderedDictType[str, T]:
    """Check that all items in the `source` iterable are of the type `type` and map them into an OrderedDict."""
    out = OrderedDict()  # type: OrderedDict[str, T]
    if source is None:
        return out

    for item in source:
        if not isinstance(item, type):
            raise TypeError(f"{item} not a {type}")
        out[getattr(item, attr)] = item
    return out


def serialize_into(
    dest,  # type: OrderedDict[str, Any] # noqa: ANN001
    key: str,
    value: Any,
    *,
    flatten_dicts: bool = False,
    elide_empty_iterables: bool = False,
) -> None:
    if value is None:
        return

    if flatten_dicts and isinstance(value, dict):
        value = list(value.values())

    if isinstance(value, (tuple, list)):  # can't use collections.Collection :(
        if elide_empty_iterables and not value:
            return
        value = [_serialize_if_able(item) for item in value]
    else:
        value = _serialize_if_able(value)

    dest[key] = value


def _serialize_if_able(v: Any) -> Any:
    if isinstance(v, Enum):
        return v.value

    return v.serialize() if hasattr(v, "serialize") else v
