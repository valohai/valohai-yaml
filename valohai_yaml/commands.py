import re
import warnings
from shlex import quote
from typing import List, Optional, TYPE_CHECKING, Union

from valohai_yaml.objs.parameter_map import ParameterMap
from valohai_yaml.utils import listify

if TYPE_CHECKING:
    from re import Match


class CommandInterpolationWarning(UserWarning):
    """Warning issued when command interpolation fails."""


interpolable_re = re.compile(r'{(.+?)}')


def quote_multiple(args: Optional[List[str]]) -> str:
    if not args:
        return ''
    return ' '.join(quote(arg) for arg in args)


def _replace_interpolation(parameter_map: ParameterMap, match: 'Match') -> str:
    value = match.group(1)
    if value in ('parameters', 'params'):
        return quote_multiple(parameter_map.build_parameters())
    elif value.startswith('parameter:'):
        parameter_name = value.split(':', 1)[1]
        if parameter_name in parameter_map.parameters:
            return quote_multiple(parameter_map.build_parameter_by_name(parameter_name))
    elif value.startswith('parameter-value:'):
        parameter_name = value.split(':', 1)[1]
        value = parameter_map.values.get(parameter_name)
        if value:
            return quote(str(value))
    return match.group(0)  # Return the original otherwise


def build_command(
    command: Union[str, List[str]],
    parameter_map: ParameterMap,
) -> List[str]:
    """
    Build command line(s) using the given parameter map.

    Even if the passed a single `command`, this function will return a list
    of shell commands.  It is the caller's responsibility to concatenate them,
    likely using the semicolon or double ampersands.

    :param command: The command to interpolate params into.
    :param parameter_map: A ParameterMap object containing parameter knowledge.

    :return: list of commands
    """
    if isinstance(parameter_map, list):
        raise TypeError("Passing in lists as ParameterMaps is no longer supported.")

    out_commands = []
    commands = listify(command)  # type: List[str]
    for command in commands:
        # Only attempt formatting if the string smells like it should be formatted.
        # This allows the user to include shell syntax in the commands, if required.
        # (There's still naturally the chance for false-positives, so guard against
        #  those value errors and warn about them.)

        if interpolable_re.search(command):
            try:
                command = interpolable_re.sub(
                    lambda match: _replace_interpolation(parameter_map, match),
                    command,
                )
            except ValueError as exc:  # pragma: no cover
                warnings.warn(
                    'failed to interpolate into %r: %s' % (command, exc),
                    CommandInterpolationWarning
                )
        out_commands.append(command.strip())
    return out_commands
