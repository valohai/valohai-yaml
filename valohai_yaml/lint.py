from typing import Any, Callable, Iterator, List, Optional

import yaml as pyyaml
from jsonschema.exceptions import relevance

from valohai_yaml.excs import ValidationError
from valohai_yaml.types import LintResultMessage, YamlReadable
from valohai_yaml.utils import read_yaml
from valohai_yaml.utils.terminal import noop_style, style
from valohai_yaml.validation import get_validator


class LintResult:
    """Container for lint results."""

    def __init__(self) -> None:
        self.messages: List[LintResultMessage] = []

    def add_error(
        self,
        message: str,
        location: None = None,
        exception: Optional[Exception] = None,
    ) -> None:
        self.messages.append(
            {
                "type": "error",
                "message": message,
                "location": location,
                "exception": exception,
            },
        )

    def add_warning(
        self,
        message: str,
        location: None = None,
        exception: Optional[Exception] = None,
    ) -> None:
        self.messages.append(
            {
                "type": "warning",
                "message": message,
                "location": location,
                "exception": exception,
            },
        )

    def add_hint(
        self,
        message: str,
        location: None = None,
        exception: Optional[Exception] = None,
    ) -> None:
        self.messages.append(
            {
                "type": "hint",
                "message": message,
                "location": location,
                "exception": exception,
            },
        )

    @property
    def warning_count(self) -> int:
        return sum(1 for m in self.messages if m["type"] == "warning")

    @property
    def error_count(self) -> int:
        return sum(1 for m in self.messages if m["type"] == "error")

    @property
    def warnings(self) -> Iterator[LintResultMessage]:
        return (m for m in self.messages if m["type"] == "warning")

    @property
    def errors(self) -> Iterator[LintResultMessage]:
        return (m for m in self.messages if m["type"] == "error")

    @property
    def hints(self) -> Iterator[LintResultMessage]:
        return (m for m in self.messages if m["type"] == "hint")

    def is_valid(self) -> bool:
        return self.warning_count == 0 and self.error_count == 0


def lint_file(file_path: str, *, validate_schema: bool = True) -> LintResult:
    """
    Validate & lint `file_path` and return a LintResult.

    :param file_path: YAML filename
    :param validate_schema: Whether to validate against the JSON schema before attempting to parse and lint.
                            This should generally always be true, but can be disabled for testing purposes.
    :return: LintResult object
    """
    with open(file_path) as yaml:
        try:
            return lint(yaml, validate_schema=validate_schema)
        except Exception as e:
            lr = LintResult()
            lr.add_error(f"could not parse YAML: {e}", exception=e)
            return lr


def _validate_json_schema(
    lr: LintResult,
    data: Any,
    styler: Callable[..., str] = style,
) -> int:
    """Validate against the JSON schema; add errors to `lr` and return the number of errors."""
    errors = sorted(
        get_validator().iter_errors(data),
        key=lambda error: (relevance(error), repr(error.path)),
    )
    for error in errors:
        simplified_schema_path = [
            el
            for el in list(error.relative_schema_path)
            if el not in ("properties", "items")
        ]
        obj_path = [str(el) for el in error.path]
        styled_validator = styler(error.validator.title(), bold=True)
        styled_schema_path = styler(".".join(simplified_schema_path), bold=True)
        styled_message = styler(error.message, fg="red")
        styled_path = styler(".".join(obj_path), bold=True)
        lr.add_error(
            f"  {styled_validator} validation on {styled_schema_path}: {styled_message} ({styled_path})",
        )
        # when path has only 2 nodes. it means it has problem in main steps/pipelines/endpoints objects
        if len(error.path) == 2 and not error.instance:
            styled_hint = styler(
                "File contains valid YAML but there might be an indentation "
                f"error in following configuration: {styled_path}",
                fg="blue",
            )
            lr.add_hint(styled_hint)
    return len(errors)


def lint(
    yaml: YamlReadable,
    *,
    validate_schema: bool = True,
    ansi_colors: bool = True,
) -> LintResult:
    """
    Validate & lint `yaml` and return a `LintResult`.

    :param yaml: YAML string or file-like object
    :param validate_schema: Whether to validate against the JSON schema before attempting to parse and lint.
                            This should generally always be true, but can be disabled for testing purposes.
    :param ansi_colors: Whether to use ANSI colors in the output.
    """
    lr = LintResult()
    try:
        data = read_yaml(yaml)
    except pyyaml.YAMLError as err:
        if hasattr(err, "problem_mark"):
            mark = err.problem_mark
            indent_error = (
                f"Indentation Error at line {mark.line + 1}, column {mark.column + 1}"
            )
            lr.add_error(indent_error)
        else:
            lr.add_error(str(err))
        return lr

    if validate_schema and _validate_json_schema(
        lr,
        data,
        styler=style if ansi_colors else noop_style,  # type: ignore[arg-type]
    ):
        # If validation found errors, don't try to parse the data
        return lr

    try:
        from valohai_yaml.objs import Config

        config = Config.parse(data)
    except ValidationError as err:  # Could happen before we get to linting things
        lr.add_error(str(err), exception=err)
    else:
        config.lint(lr, context={})
    return lr
