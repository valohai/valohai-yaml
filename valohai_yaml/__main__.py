import argparse
import sys
from typing import List, Optional

from valohai_yaml.lint import lint, LintResult


def main(argv: Optional[List[str]] = None) -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument(
        '--strict-warnings',
        action='store_true',
        help='exit with error if there are warnings',
    )
    ap.add_argument('file', nargs='+', help='file(s) to validate')
    args = ap.parse_args(argv)
    errors = warnings = 0
    for file in args.file:
        result = process_file(file)
        errors += result.error_count
        warnings += result.warning_count

    if errors or warnings:
        print('*** {} errors, {} warnings'.format(errors, warnings))

    return 1 if errors or (args.strict_warnings and warnings) else 0


def process_file(file: str) -> LintResult:
    header_printed = False

    with open(file, 'rb') as stream:
        result = lint(stream)

    for item in result.messages:
        if not header_printed:
            print('>>>', file)
            header_printed = True
        print('{}: {}'.format(item['type'], item['message']))
        print('-' * 60)
    return result


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover
