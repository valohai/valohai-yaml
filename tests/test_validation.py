import glob
import os

import pytest

from tests.consts import (
    error_examples_path,
    examples_path,
    invalid_obj,
    valid_bytes,
    valid_obj,
    warning_examples_path,
)
from tests.utils import get_error_example_path, get_valid_example_path
from valohai_yaml import ValidationErrors, validate
from valohai_yaml.__main__ import main


@pytest.mark.parametrize(
    "good_example_path",
    glob.glob(os.path.join(examples_path, "*.yaml")),
)
def test_good_examples_cli(good_example_path):
    """Test that known-good files validate via the CLI."""
    assert main([good_example_path]) == 0


@pytest.mark.parametrize(
    "bad_example_path",
    glob.glob(os.path.join(error_examples_path, "*.yaml")),
)
def test_bad_examples_cli(capsys, bad_example_path):
    """Test that bad examples don't validate via the CLI."""
    assert main([bad_example_path]) == 1
    out, err = capsys.readouterr()
    assert out


@pytest.mark.parametrize(
    "yaml_path",
    glob.glob(os.path.join(warning_examples_path, "*.yaml")),
)
def test_warning_examples_cli(capsys, yaml_path):
    """Test that warning examples don't validate via the CLI in strict mode."""
    assert main(["--strict-warnings", yaml_path]) == 1
    out, err = capsys.readouterr()
    assert out


def test_invalid_file_missing_properties_cli(capsys):
    assert main([get_error_example_path("step-missing-required-properties.yaml")]) == 1
    out, err = capsys.readouterr()
    assert "step-missing-required-properties.yaml" in out
    assert "'command' is a required property" in out
    assert "'name' is a required property" in out
    assert "8 is not valid under any of the given schemas" in out


def test_invalid_file_too_long_input_name_cli(capsys):
    assert main([get_error_example_path("input-name-too-long.yaml")]) == 1
    out, err = capsys.readouterr()
    assert "input-name-too-long.yaml" in out
    assert (
        "'this-input-name-is-way-too-long-and-will-cause-the-validation-to-fail' is too long"
        in out
    )


def test_invalid_on_error_value(capsys):
    assert main([get_error_example_path("invalid-on-error-value.yaml")]) == 1
    out, err = capsys.readouterr()
    assert "invalid-on-error-value.yaml" in out
    assert "is not valid under any of the given schemas" in out


def test_valid_endpoint_name_cli(capsys):
    assert main([get_valid_example_path("endpoint-names-valid.yaml")]) == 0


def test_invalid_endpoint_name_cli(capsys):
    assert main([get_error_example_path("endpoint-names-invalid.yaml")]) == 1
    out, err = capsys.readouterr()
    assert "endpoint-names-invalid.yaml" in out
    assert "'wsgi_endpoint' does not match '^[a-z][a-z0-9-]+$'" in out
    assert "'server@endpoint' does not match '^[a-z][a-z0-9-]+$'" in out
    assert "'3-server-endpoint' does not match '^[a-z][a-z0-9-]+$'" in out
    assert "'Server-endpoint' does not match '^[a-z][a-z0-9-]+$'" in out


def test_bytes_validation():
    """Test that you can pass in bytestring YAML and validate it."""
    assert not validate(valid_bytes)


def test_popo_validation():
    """Test that you can pass in pre-parsed data and validate it."""
    # (POPO: Plain Old Python Object.)
    assert not validate(valid_obj)


def test_raise():
    with pytest.raises(ValidationErrors) as ei:
        validate(invalid_obj, raise_exc=True)
    assert list(ei.value)  # test iteration over errors


def test_error_list():
    errs = [f"{err}" for err in validate(invalid_obj, raise_exc=False)]
    assert any(
        ("Additional properties are not allowed" in err) for err in errs
    )  # pragma: no branch
    assert any(("required property" in err) for err in errs)  # pragma: no branch
    assert any(("0 is not of type 'string'" in err) for err in errs)
