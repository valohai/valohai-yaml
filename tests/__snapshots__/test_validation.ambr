# serializer version: 1
# name: test_bad_examples_cli[endpoint-names-invalid.yaml]
  '''
  >>> endpoint-names-invalid.yaml
  error:   Pattern validation on endpoint.name.pattern: 'Server-endpoint' does not match '^[a-z][a-z0-9-]+$' (0.endpoint.name)
  ------------------------------------------------------------
  error:   Pattern validation on endpoint.name.pattern: '3-server-endpoint' does not match '^[a-z][a-z0-9-]+$' (1.endpoint.name)
  ------------------------------------------------------------
  error:   Pattern validation on endpoint.name.pattern: 'server@endpoint' does not match '^[a-z][a-z0-9-]+$' (2.endpoint.name)
  ------------------------------------------------------------
  error:   Pattern validation on endpoint.name.pattern: 'wsgi_endpoint' does not match '^[a-z][a-z0-9-]+$' (3.endpoint.name)
  ------------------------------------------------------------
  *** 4 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[input-name-too-long.yaml]
  '''
  >>> input-name-too-long.yaml
  error:   Maxlength validation on step.inputs.name.maxLength: 'this-input-name-is-way-too-long-and-will-cause-the-validation-to-fail' is too long (0.step.inputs.0.name)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-YAML-indentation.yaml]
  '''
  >>> invalid-YAML-indentation.yaml
  error: Indentation Error at line 3, column 10
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-edge-parts.yaml]
  '''
  >>> invalid-edge-parts.yaml
  error: Target specifier 'batch2.input' must have 3 parts (it has 2)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-indentation-with-valid-YAML.yaml]
  '''
  >>> invalid-indentation-with-valid-YAML.yaml
  error:   Type validation on step.type: None is not of type 'object' (0.step)
  ------------------------------------------------------------
  hint: File contains valid YAML but there might be an indentation error in following configuration: 0.step
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-on-error-value.yaml]
  '''
  >>> invalid-on-error-value.yaml
  error:   Anyof validation on pipeline.nodes.anyOf: {'name': 'train-with-errors', 'step': 'train', 'type': 'task', 'on-error': 'report-to-sentry'} is not valid under any of the given schemas (2.pipeline.nodes.1)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-pipeline-with-override-example.yaml]
  '''
  >>> invalid-pipeline-with-override-example.yaml
  error: Pipeline My overriden input pipeline node merged step base step: input training-labels-error does not exist in step
  ------------------------------------------------------------
  error: Pipeline My overriden input pipeline node merged step base step: parameter d does not exist in step
  ------------------------------------------------------------
  error: Pipeline My overriden input pipeline node overridden step secon step does not exist
  ------------------------------------------------------------
  *** 3 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-shorthand-edge.yaml]
  '''
  >>> invalid-shorthand-edge.yaml
  error: Malformed edge shorthand ['batch1.input.training-labels*']
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[invalid-upload-store-value.yaml]
  '''
  >>> invalid-upload-store-value.yaml
  error:   Maxlength validation on step.upload-store.maxLength: 'This is a test string exactly with  65 characters long in length.' is too long (0.step.upload-store)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[step-invalid-resources.yaml]
  '''
  >>> step-invalid-resources.yaml
  error:   Type validation on step.resources.cpu.min.type: 'Not a number' is not of type 'number' (0.step.resources.cpu.min)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[step-missing-required-properties.yaml]
  '''
  >>> step-missing-required-properties.yaml
  error:   Oneof validation on step.command.oneOf: 8 is not valid under any of the given schemas (1.step.command)
  ------------------------------------------------------------
  error:   Type validation on step.image.type: True is not of type 'string' (0.step.image)
  ------------------------------------------------------------
  error:   Required validation on step.required: 'command' is a required property (0.step)
  ------------------------------------------------------------
  error:   Required validation on step.required: 'name' is a required property (0.step)
  ------------------------------------------------------------
  *** 4 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[step-stop-condition.yaml]
  '''
  >>> step-stop-condition.yaml
  error: Step no-stop, `stop-condition` is not a valid expression: invalid syntax (<expression>, line 1)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[step-with-error-tasks.yaml]
  '''
  >>> step-with-error-tasks.yaml
  error:   Type validation on task.parameters.rules.type: None is not of type 'object' (1.task.parameters.0.rules)
  ------------------------------------------------------------
  error:   Type validation on task.parameters.name.type: None is not of type 'string' (2.task.parameters.0.name)
  ------------------------------------------------------------
  error:   Required validation on task.parameters.required: 'style' is a required property (2.task.parameters.0)
  ------------------------------------------------------------
  error:   Required validation on task.parameters.required: 'name' is a required property (5.task.parameters.0)
  ------------------------------------------------------------
  error:   Required validation on task.parameters.required: 'style' is a required property (5.task.parameters.0)
  ------------------------------------------------------------
  error:   Required validation on task.required: 'name' is a required property (1.task)
  ------------------------------------------------------------
  error:   Required validation on task.required: 'step' is a required property (4.task)
  ------------------------------------------------------------
  *** 7 errors, 0 warnings
  
  '''
# ---
# name: test_bad_examples_cli[task-stop-condition.yaml]
  '''
  >>> task-stop-condition.yaml
  error: Task no-stop, `stop-condition` is not a valid expression: EOL while scanning string literal (<expression>, line 1)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
# name: test_warning_examples_cli[invalid-numeric-parameter-default.yaml]
  '''
  >>> invalid-numeric-parameter-default.yaml
  warning: Step test, parameter value: default zoom is not an integer
  ------------------------------------------------------------
  *** 0 errors, 1 warnings
  
  '''
# ---
# name: test_warning_examples_cli[invalid-numeric-range-parameter-default.yaml]
  '''
  >>> invalid-numeric-range-parameter-default.yaml
  warning: Step test, parameter value: default 1337 is greater than the maximum allowed (100)
  ------------------------------------------------------------
  *** 0 errors, 1 warnings
  
  '''
# ---
# name: test_warning_examples_cli[invalid-string-parameter-default.yaml]
  '''
  >>> invalid-string-parameter-default.yaml
  warning: Step test, parameter ctime: default `default` value datetime.datetime(2020, 9, 2, 0, 0, tzinfo=datetime.timezone.utc) is not a string (got a <class 'datetime.datetime'>)
  ------------------------------------------------------------
  *** 0 errors, 1 warnings
  
  '''
# ---
# name: test_warning_examples_cli[nonsensical-task-bits.yaml]
  '''
  >>> nonsensical-task-bits.yaml
  warning: optimization_target_metric only makes sense for Bayesian TPE tasks
  ------------------------------------------------------------
  warning: optimization_target_value only makes sense for Bayesian TPE tasks
  ------------------------------------------------------------
  *** 0 errors, 2 warnings
  
  '''
# ---
# name: test_warning_examples_cli[override-with-extra-fields-warning.yaml]
  '''
  >>> override-with-extra-fields-warning.yaml
  error:   Anyof validation on pipeline.nodes.anyOf: {'name': 'merged', 'type': 'execution', 'step': 'base step', 'override': {'training-images': 's3://hello/hello.jpg'}} is not valid under any of the given schemas (1.pipeline.nodes.0)
  ------------------------------------------------------------
  *** 1 errors, 0 warnings
  
  '''
# ---
